src/csexp_rpc/io_buffer.ml

```diff
@@ -0,0 +1,79 @@
+open Stdune
+
+type t =
+  { mutable bytes : Bytes.t (* underlying bytes *)
+  ; (* the position we can start reading from (until [pos_w]) *)
+    mutable pos_r : int
+  ; (* the position we can start writing to (until [Bytes.length bytes - 1]) *)
+    mutable pos_w : int
+  ; (* total number of bytes written to this buffer. 2^63 bytes should be
+       enough for anybody *)
+    mutable total_written : int
+  }
+
+type flush_token = int
+
+(* We can't use [Out_channel] for writes on Linux because we want to disable
+   sigpipes. Eventually we'll move to event based IO and ditch the threads,
+   so we'll need this anyway *)
+
+let create () =
+  { bytes = Bytes.create 8192; pos_r = 0; pos_w = 0; total_written = 0 }
+
+let maybe_resize_to_fit t write_size =
+  let buf_len = Bytes.length t.bytes in
+  let capacity = buf_len - t.pos_w in
+  if capacity < write_size then (
```

Comment:
Often it is better to grow the buffer exponentially (eg by multiplying the length by a power of 2) to keep performing a linear number of resizings.